datalad/distribution/publish.py

```diff
@@ -192,48 +210,60 @@ def __call__(dataset=None, dest=None, path=None,
-                                # use their default as well
-                                with_data=with_data,
-                                recursive=recursive))
-                return results
-
-            return ds
-
-        elif exists(path):
-            # At this point `path` is not referencing a (sub)dataset.
-            # An annexed file is the only thing left, that `path` might be
-            # validly pointing to. Anything else we can't handle currently.
-            if isinstance(ds.repo, AnnexRepo):
-                try:
-                    if ds.repo.get_file_key(relativepath):
-                        # file is in annex, publish it
-                        ds.repo._run_annex_command('copy',
-                                                   annex_options=[path,
-                                                                  '--to=%s' % dest_resolved])
-                        return path
-                except (FileInGitError, FileNotInAnnexError):
-                    pass
-            # `path` can't be published
-            lgr.warning("Don't know how to publish %s." % path)
-            return None
+            # we need to fetch
+            # TODO
+            # Note: This is about a gitpython issue as well as something about
+            # annex -> might mean, that we need to do it in case we pushed an annex branch only.
+            # Apparently, we can annex copy new files only, after this fetch. Figure it out!
+            ds.repo.fetch(remote=dest_resolved)
+
+            results.append(ds)
+
+            if publish_files or to_annex:
+                if not isinstance(ds.repo, AnnexRepo):
+                    raise RuntimeError(
+                        "Cannot publish content of something, that is not "
+                        "part of an annex. ({0})".format(ds))
+
+                lgr.info("Publishing data of dataset {0} ...".format(ds))
+                results += ds.repo.copy_to(files=publish_files,
+                                           remote=dest_resolved,
+                                           options=to_annex)
+
+        for dspath in expl_subs:
+            # these datasets need to be pushed regardless of additional paths
+            # pointing inside them
+            # due to API, this may not happen when calling publish with paths,
+            # therefore force it.
+            # TODO: There might be a better solution to avoid two calls of
+            # publish() on the very same Dataset instance
+            results += Dataset(opj(ds.path, dspath)).publish(
+                to=to, recursive=recursive)
+
+        for d in publish_subs:
+            # recurse into subdatasets
+            results += publish_subs[d]['dataset'].publish(
+                to=to,
+                path=publish_subs[d]['files'],
+                recursive=recursive,
+                to_annex=to_annex)
+
+        return results
+
+    @staticmethod
+    def result_renderer_cmdline(res):
```

Comment:
I thought that there is already something mocking UI away, allowing also for testing....  see `@with_testsui`